aiida/tools/visualization/graph.py

```diff
@@ -0,0 +1,685 @@
+from aiida.orm.querybuilder import QueryBuilder
+from aiida.plugins import BaseFactory
+from aiida.common import LinkType
+
+
+def default_link_styles(link_type):
+    """map link_type to a graphviz edge style
+
+    :param link_type: a LinkType attribute
+    :type link_type: One of the members from
+        :class:`aiida.common.links.LinkType`
+    :rtype: dict
+
+    """
+    return {
+        LinkType.INPUT_CALC: {
+            "style": "solid",
+            "color": "#000000"  # black
+        },
+        LinkType.INPUT_WORK: {
+            "style": "dashed",
+            "color": "#0000FF"  # blue
+        },
+        LinkType.CALL_CALC: {
+            "style": "dotted",
+            "color": "#000000"  # black
+        },
+        LinkType.CALL_WORK: {
+            "style": "dotted",
+            "color": "#0000FF"  # blue
+        },
+        LinkType.CREATE: {
+            "style": "solid",
+            "color": "#006400"  # green
+        },
+        LinkType.RETURN: {
+            "style": "dashed",
+            "color": "#006400"  # green
+        }
+    }[link_type]
+
+
+def default_data_styles(node):
+    """map a data node to a graphviz node style
+
+    :param node: the node to map
+    :type node: aiida.orm.nodes.node.Node
+    :rtype: dict
+
+    """
+    class_node_type = node.class_node_type
+    if hasattr(node, "get_style_default"):
+        default = node.get_style_default()
+    else:
+        default = {
+            "shape": "ellipse",
+            "style": "filled",
+            "fillcolor": "#FFFFF0",  # ivory
+            "pencolor": "black"
+        }
+    mapping = {
+        "data.code.Code.": {
+            "shape": "diamond",
+            "style": "filled",
+            "fillcolor": "#FFFFFF",  # white
+            "pencolor": "orange"
+        }
+    }
+    return mapping.get(class_node_type, default)
+
+
+def default_data_sublabels(node):
+    """function mapping data nodes to a sublabel
+    (e.g. specifying some attribute values)
+
+    :param node: the node to map
+    :type node: aiida.orm.nodes.node.Node
+    :rtype: str
+
+    """
+    class_node_type = node.class_node_type
+    if class_node_type == "data.int.Int.":
+        sublabel = "value: {}".format(node.get_attribute("value", ""))
+    elif class_node_type == "data.float.Float.":
+        sublabel = "value: {}".format(node.get_attribute("value", ""))
+    elif class_node_type == "data.code.Code.":
+        sublabel = "{}@{}".format(os.path.basename(node.get_execname()), node.get_computer_name())
+    elif class_node_type == "data.singlefile.SinglefileData.":
+        sublabel = node.filename
+    elif class_node_type == "data.remote.RemoteData.":
+        sublabel = "@{}".format(node.get_computer_name())
+    elif class_node_type == "data.structure.StructureData.":
+        sublabel = node.get_formula()
+    elif class_node_type == "data.cif.CifData.":
+        formulae = [str(f).replace(" ", "") for f in node.get_attribute('formulae', []) if f]
```

Comment:
This gives an error, when `formulae` is explicitly set to `None`, e.g.
```
In [1]: n = Dict()

In [2]: n.get_attribute('test', '1')
Out[2]: '1'

In [3]: n.get_attribute('test', [])
Out[3]: []

In [4]: n.set_attribute('test', None)

In [5]: n.get_attribute('test', [])  # returns None
```
This is the same behavior as with `dict` and `get`.

One could argue that `CifData` *should* not have `formulae` set to `None` but they exist in the wild and so we need to deal with them.

The solution would be simply
```
formulae = [str(f).replace(" ", "") for f in node.get_attribute('formulae') or [] if f]
```
(same for next line)

Reply:
Implemented in https://github.com/aiidateam/aiida_core/pull/2596/commits/2c0899046f6b20065777b9ef325e2b2de8b0e075